cf-custom-resources/lib/custom-domain-app-runner.js

```diff
@@ -0,0 +1,304 @@
+
+        if (domain.Status !== DOMAIN_STATUS_ACTIVE) {
+            // Exponential backoff with jitter based on 200ms base
+            // component of backoff fixed to ensure minimum total wait time on
+            // slow targets.
+            const base = Math.pow(2, i);
+            await sleep(Math.random() * base * 50 + base * 150);
+            continue;
+        }
+        return;
+    }
+
+    if (i === ATTEMPTS_WAIT_FOR_ACTIVE) {
+        console.log("Fail to wait for state to become ACTIVE. However, this doesn't necessarily mean the operation has failed. It usually takes a long time to validate domain and can be longer than the 15 minutes duration for which a Lambda function can run at most.");
+        throw new Error(`fail to wait for domain ${customDomainName} to become ${DOMAIN_STATUS_ACTIVE}`);
+    }
+}
+
+/**
+ * Validate certificates of the custom domain for the service by upserting validation records.
+ * Errors are not handled and are directly passed to the caller.
+ *
+ * @param {string} serviceARN ARN of the service that the custom domain applies to.
+ * @param {string} customDomainName the custom domain name.
+ */
+async function addCustomDomain(serviceARN, customDomainName) {
+    const data = await appRunnerClient.associateCustomDomain({
+        DomainName: customDomainName,
+        ServiceArn: serviceARN,
+    }).promise();
+
+    await Promise.all([
+        updateCNAMERecordAndWait(customDomainName, data.DNSTarget, appHostedZoneID, "UPSERT"), // Upsert the record that maps `customDomainName` to the DNS of the app runner service.
+        validateCertForDomain(serviceARN, customDomainName),
+    ]);
+}
+
+/**
+ * Validate certificates of the custom domain for the service by upserting validation records.
+ *
+ * @param {string} serviceARN ARN of the service that the custom domain applies to.
+ * @param {string} domainName the custom domain name.
+ * @throws wrapped error.
+ */
+async function validateCertForDomain(serviceARN, domainName) {
+    let i;
+    for (i = 0; i < ATTEMPTS_WAIT_FOR_PENDING; i++){
+        const data = await appRunnerClient.describeCustomDomains({
+            ServiceArn: serviceARN,
+        }).promise().catch(err => {
+            throw new Error(`update validation records for domain ${domainName}: ` + err.message);
+        });
+
+        let domain;
+        for (const d of data.CustomDomains) {
+            if (d.DomainName === domainName) {
+                domain = d;
+                break;
+            }
+        }
+
+        if (!domain) {
+            throw new Error(`update validation records for domain ${domainName}: domain ${domainName} is not associated`);
+        }
+
+        if (domain.Status !== DOMAIN_STATUS_PENDING_VERIFICATION) {
+            await sleep(3000);
+            continue;
+        }
+
+        // Upsert all records needed for certificate validation.
+        const records = domain.CertificateValidationRecords;
+        for (const record of records) {
```

Comment:
right now we would be making this call multiple times it the domain.Status is stuck on: `PENDING_CERTIFICATE_DNS_VALIDATION` should we be ensuring that we make the request only once?